File: src/main/scala/resonantinduction/archaic/firebox/TileHotPlate.java
package resonantinduction.archaic.firebox;

import cpw.mods.fml.common.network.ByteBufUtils;
import cpw.mods.fml.relauncher.Side;
import cpw.mods.fml.relauncher.SideOnly;
import io.netty.buffer.ByteBuf;
import net.minecraft.block.material.Material;
import net.minecraft.client.renderer.texture.IIconRegister;
import net.minecraft.entity.player.EntityPlayer;
import net.minecraft.item.ItemStack;
import net.minecraft.item.crafting.FurnaceRecipes;
import net.minecraft.nbt.NBTTagCompound;
import net.minecraft.tileentity.TileEntity;
import net.minecraft.util.IIcon;
import resonant.content.spatial.block.SpatialBlock;
import resonant.lib.content.prefab.java.TileInventory;
import resonant.lib.network.discriminator.PacketTile;
import resonant.lib.network.discriminator.PacketType;
import resonant.lib.network.handle.IPacketReceiver;
import resonantinduction.core.Reference;
import universalelectricity.core.transform.region.Cuboid;
import universalelectricity.core.transform.vector.Vector2;
import universalelectricity.core.transform.vector.Vector3;

import java.util.ArrayList;
import java.util.List;

/**
 * For smelting items.
 *
 * @author Calclavia
 */
public class TileHotPlate extends TileInventory implements IPacketReceiver
{
	public static final int MAX_SMELT_TIME = 200;
	public final int[] smeltTime = new int[] { 0, 0, 0, 0 };
	public final int[] stackSizeCache = new int[] { 0, 0, 0, 0 };
	private final int POWER = 50000;

	public TileHotPlate()
	{
		super(Material.iron);
		setSizeInventory(4);
		bounds(new Cuboid(0, 0, 0, 1, 0.2f, 1));
		normalRender(false);
		forceStandardRender(true);
		isOpaqueCube(false);
	}

	@Override
	public void update()
	{
		if (canRun())
		{
			boolean didSmelt = false;

			for (int i = 0; i < getSizeInventory(); i++)
			{
				if (canSmelt(this.getStackInSlot(i)))
				{
					if (smeltTime[i] <= 0)
					{
						/**
						 * Heat up all slots
						 */
						stackSizeCache[i] = this.getStackInSlot(i).stackSize;
						smeltTime[i] = MAX_SMELT_TIME * stackSizeCache[i];
						worldObj.markBlockForUpdate(xCoord, yCoord, zCoord);
					}
					else if (smeltTime[i] > 0)
					{
						/**
						 * Do the smelt action.
						 */
						if (--smeltTime[i] == 0)
						{
							if (!worldObj.isRemote)
							{
								ItemStack outputStack = FurnaceRecipes.smelting().getSmeltingResult(getStackInSlot(i)).copy();
								outputStack.stackSize = stackSizeCache[i];
								setInventorySlotContents(i, outputStack);
								worldObj.markBlockForUpdate(xCoord, yCoord, zCoord);
							}
						}
					}

					didSmelt = true;
				}
				else
				{
					smeltTime[i] = 0;
				}
			}
		}
	}

	public void onInventoryChanged()
	{
		//super.onInventoryChanged();

		/**
		 * Update cache calculation.
		 */
		for (int i = 0; i < getSizeInventory(); i++)
		{
			if (getStackInSlot(i) != null)
			{
				if (stackSizeCache[i] != getStackInSlot(i).stackSize)
				{
					if (smeltTime[i] > 0)
					{
						smeltTime[i] += (getStackInSlot(i).stackSize - stackSizeCache[i]) * MAX_SMELT_TIME;
					}

					stackSizeCache[i] = getStackInSlot(i).stackSize;
				}
			}
			else
			{
				stackSizeCache[i] = 0;
			}
		}

		if (worldObj != null)
		{
			worldObj.markBlockForUpdate(xCoord, yCoord, zCoord);
		}
	}

	public boolean canRun()
	{

		TileEntity tileEntity = worldObj.getTileEntity(xCoord, yCoord - 1, zCoord);

		if (tileEntity instanceof TileFirebox)
		{
			if (((TileFirebox) tileEntity).isBurning())
			{
				return true;
			}
		}

		return false;
	}

	public boolean canSmelt(ItemStack stack)
	{
		return FurnaceRecipes.smelting().getSmeltingResult(stack) != null;
	}

	public boolean isSmelting()
	{
		for (int i = 0; i < getSizeInventory(); i++)
		{
			if (getSmeltTime(i) > 0)
			{
				return true;
			}
		}

		return false;
	}

	public int getSmeltTime(int i)
	{
		return smeltTime[i];
	}

	@Override
	public boolean isItemValidForSlot(int i, ItemStack itemStack)
	{
		return i < getSizeInventory() && canSmelt(itemStack);
	}

	@Override
	public PacketTile getDescPacket()
	{
		return new PacketTile(this, this.getPacketData(0).toArray());
	}

	/**
	 * 1 - Description Packet
	 * 2 - Energy Update
	 * 3 - Tesla Beam
	 */
	public List getPacketData(int type)
	{
		List list = new ArrayList();
		NBTTagCompound nbt = new NBTTagCompound();
		this.writeToNBT(nbt);
		list.add(nbt);
		return list;
	}

	@Override
	public void read(ByteBuf data, EntityPlayer player, PacketType type)
	{
		try
		{
			this.readFromNBT(ByteBufUtils.readTag(data));
			this.worldObj.markBlockForUpdate(this.xCoord, this.yCoord, this.zCoord);
		}
		catch (Exception e)
		{
			e.printStackTrace();
		}
	}

	@Override
	public void readFromNBT(NBTTagCompound nbt)
	{
		super.readFromNBT(nbt);
		for (int i = 0; i < getSizeInventory(); i++)
		{
			smeltTime[i] = nbt.getInteger("smeltTime" + i);
		}
	}

	@Override
	public void writeToNBT(NBTTagCompound nbt)
	{
		super.writeToNBT(nbt);
		for (int i = 0; i < getSizeInventory(); i++)
		{
			nbt.setInteger("smeltTime" + i, smeltTime[i]);
		}
	}

	@Override
	@SideOnly(Side.CLIENT)
	public void registerIcons(IIconRegister iconReg)
	{
		super.registerIcons(iconReg);
		SpatialBlock.icon().put("electricHotPlate", iconReg.registerIcon(Reference.prefix() + "electricHotPlate"));
	}

	@Override
	@SideOnly(Side.CLIENT)
	public IIcon getIcon(int side, int meta)
	{
		return meta == 1 ? SpatialBlock.icon().get("electricHotPlate") : SpatialBlock.icon().get(getTextureName());
	}

	@Override
	public void click(EntityPlayer player)
	{
		if (server())
		{
			extractItem(this, 0, player);
		}
	}

	@Override
	public boolean use(EntityPlayer player, int side, Vector3 hit)
	{
		if (server())
		{
			Vector2 hitVector = new Vector2(hit.x(), hit.z());
			final double regionLength = 1d / 2d;

			/**
			 * Crafting Matrix
			 */
			matrix:
			for (int j = 0; j < 2; j++)
			{
				for (int k = 0; k < 2; k++)
				{
					Vector2 check = new Vector2(j, k).multiply(regionLength);

					if (check.distance(hitVector) < regionLength)
					{
						int slotID = j * 2 + k;
						interactCurrentItem(this, slotID, player);
						break matrix;
					}
				}
			}

			onInventoryChanged();
		}

		return true;
	}
}

File: src/main/scala/resonantinduction/archaic/firebox/TileHotPlate.scala
package resonantinduction.archaic.firebox

import cpw.mods.fml.relauncher.{Side, SideOnly}
import io.netty.buffer.ByteBuf
import net.minecraft.block.material.Material
import net.minecraft.client.renderer.texture.IIconRegister
import net.minecraft.entity.player.EntityPlayer
import net.minecraft.item.ItemStack
import net.minecraft.item.crafting.FurnaceRecipes
import net.minecraft.nbt.NBTTagCompound
import net.minecraft.util.IIcon
import net.minecraft.world.IBlockAccess
import resonant.lib.content.prefab.TInventory
import resonant.lib.network.discriminator.PacketType
import resonant.lib.network.handle.{TPacketReceiver, TPacketSender}
import resonant.lib.prefab.tile.spatial.{SpatialBlock, SpatialTile}
import resonant.lib.transform.region.Cuboid
import resonant.lib.transform.vector.{Vector2, Vector3}
import resonant.lib.wrapper.ByteBufWrapper._
import resonant.lib.wrapper.RandomWrapper._
import resonantinduction.core.Reference

/**
 * For smelting items.
 *
 * @author Calclavia
 */
object TileHotPlate
{
  final val maxSmeltTime: Int = 200
}

class TileHotPlate extends SpatialTile(Material.iron) with TInventory with TPacketSender with TPacketReceiver
{
  /** Amount of smelt time left */
  final val smeltTime = Array[Int](0, 0, 0, 0)
  final val stackSizeCache = Array[Int](0, 0, 0, 0)

  //Constructor
  bounds = new Cuboid(0, 0, 0, 1, 0.2f, 1)
  forceItemToRenderAsBlock = true
  isOpaqueCube = false
  tickRandomly = true

  override def update()
  {
    super.update()

    if (canRun)
    {
      var didSmelt = false

      for (i <- 0 until getSizeInventory)
      {
        if (canSmelt(this.getStackInSlot(i)))
        {
          if (smeltTime(i) <= 0)
          {
            stackSizeCache(i) = this.getStackInSlot(i).stackSize
            smeltTime(i) = TileHotPlate.maxSmeltTime * stackSizeCache(i)
            worldObj.markBlockForUpdate(xCoord, yCoord, zCoord)
          }
          else if (smeltTime(i) > 0)
          {
            smeltTime(i) -= 1

            if (smeltTime(i) == 0)
            {
              if (!worldObj.isRemote)
              {
                val outputStack = FurnaceRecipes.smelting.getSmeltingResult(getStackInSlot(i)).copy
                outputStack.stackSize = stackSizeCache(i)
                setInventorySlotContents(i, outputStack)
                markUpdate()
              }
            }
          }
          didSmelt = true
        }
        else
        {
          smeltTime(i) = 0
        }
      }
    }
  }

  def canRun: Boolean =
  {
    val tileEntity = worldObj.getTileEntity(xCoord, yCoord - 1, zCoord)

    if (tileEntity.isInstanceOf[TileFirebox])
    {
      if ((tileEntity.asInstanceOf[TileFirebox]).isBurning)
      {
        return true
      }
    }
    return false
  }

  override def randomDisplayTick()
  {
    val height = 0.2
    val deviation = 0.22

    (0 to 3).foreach(
      i =>
      {
        if (smeltTime(i) > 0)
        {
          for (t <- 0 until (TileHotPlate.maxSmeltTime * stackSizeCache(i) - smeltTime(i)))
          {
            i match
            {
              case 0 => world.spawnParticle("smoke", x + 0.5 - deviation + world.rand.deviate(0.1), y + height, z + 0.5 - deviation + world.rand.deviate(0.1), 0.0D, 0.0D, 0.0D)
              case 1 => world.spawnParticle("smoke", x + 0.5 - deviation + world.rand.deviate(0.1), y + height, z + 0.5 + deviation + world.rand.deviate(0.1), 0.0D, 0.0D, 0.0D)
              case 2 => world.spawnParticle("smoke", x + 0.5 + deviation + world.rand.deviate(0.1), y + height, z + 0.5 - deviation + world.rand.deviate(0.1), 0.0D, 0.0D, 0.0D)
              case 3 => world.spawnParticle("smoke", x + 0.5 + deviation + world.rand.deviate(0.1), y + height, z + 0.5 + deviation + world.rand.deviate(0.1), 0.0D, 0.0D, 0.0D)
            }
          }
        }
      }
    )
  }

  def isSmelting: Boolean = (0 until getSizeInventory).exists(getSmeltTime(_) > 0)

  def getSmeltTime(i: Int): Int =
  {
    return smeltTime(i)
  }

  override def isItemValidForSlot(i: Int, itemStack: ItemStack): Boolean =
  {
    return i < getSizeInventory && canSmelt(itemStack)
  }

  def canSmelt(stack: ItemStack): Boolean = stack != null && FurnaceRecipes.smelting.getSmeltingResult(stack) != null

  override def write(buf: ByteBuf, id: Int)
  {
    super.write(buf, id)
    val nbt = new NBTTagCompound
    writeToNBT(nbt)
    buf <<< nbt
  }

  override def writeToNBT(nbt: NBTTagCompound)
  {
    super.writeToNBT(nbt)

    (0 until stackSizeCache.size).foreach(i => nbt.setInteger("stackSizeCache" + i, stackSizeCache(i)))
    (0 until getSizeInventory).foreach(i => nbt.setInteger("smeltTime" + i, smeltTime(i)))
  }

  override def getSizeInventory: Int = 4

  override def read(buf: ByteBuf, id: Int, packetType: PacketType)
  {
    super.read(buf, id, packetType)

    readFromNBT(buf.readTag())
    markRender()
  }

  override def readFromNBT(nbt: NBTTagCompound)
  {
    super.readFromNBT(nbt)

    (0 until stackSizeCache.size).foreach(i => stackSizeCache(i) = nbt.getInteger("stackSizeCache" + i))
    (0 until getSizeInventory).foreach(i => smeltTime(i) = nbt.getInteger("smeltTime" + i))

  }

  @SideOnly(Side.CLIENT)
  override def registerIcons(iconReg: IIconRegister)
  {
    super.registerIcons(iconReg)
    SpatialBlock.icon.put("electricHotPlate", iconReg.registerIcon(Reference.prefix + "electricHotPlate"))
    SpatialBlock.icon.put("hotPlate_on", iconReg.registerIcon(Reference.prefix + "hotPlate_on"))
  }

  /**
   * Called in the world.
   */
  override def getIcon(access: IBlockAccess, side: Int): IIcon =
  {
    return if (access.getBlockMetadata(xi, yi, zi) == 1) SpatialBlock.icon.get("electricHotPlate") else (if (canRun) SpatialBlock.icon.get("hotPlate_on") else SpatialBlock.icon.get(getTextureName))
  }

  @SideOnly(Side.CLIENT)
  override def getIcon(side: Int, meta: Int): IIcon =
  {
    return if (meta == 1) SpatialBlock.icon.get("electricHotPlate") else SpatialBlock.icon.get(getTextureName)
  }

  override def click(player: EntityPlayer)
  {
    if (server)
    {
      extractItem(this, 0, player)
    }
  }

  override def use(player: EntityPlayer, side: Int, hit: Vector3): Boolean =
  {
    if (server)
    {
      val hitVector: Vector2 = new Vector2(hit.x, hit.z)
      val regionLength: Double = 1d / 2d
      var j: Int = 0
      for (j <- 0 until 2)
      {
        for (k <- 0 until 2)
        {
          val check = new Vector2(j, k) * regionLength
          if (check.distance(hitVector) < regionLength)
          {
            val slotID = j * 2 + k
            interactCurrentItem(this, slotID, player)
            onInventoryChanged()
            return true
          }
        }
      }
      return false
    }
    return true
  }

  override def onInventoryChanged()
  {
    for (i <- 0 until getSizeInventory)
    {
      if (getStackInSlot(i) != null)
      {
        if (stackSizeCache(i) != getStackInSlot(i).stackSize)
        {
          if (smeltTime(i) > 0)
          {
            smeltTime(i) += (getStackInSlot(i).stackSize - stackSizeCache(i)) * TileHotPlate.maxSmeltTime
          }
          stackSizeCache(i) = getStackInSlot(i).stackSize
        }
      }
      else
      {
        stackSizeCache(i) = 0
      }
    }
    if (worldObj != null)
    {
      worldObj.markBlockForUpdate(xCoord, yCoord, zCoord)
    }
  }

}